src/components/include/sl.h

```diff
@@ -382,13 +404,24 @@ sl_cs_exit_schedule_nospin_arg(struct sl_thd *to)
 		if (unlikely(!pt)) t = sl__globals()->idle_thd;
 		else               t = sl_mod_thd_get(pt);
 	}
-	thdcap = t->thdcap;
-	prio   = t->prio;
 
+	if (t->flags & SL_FLAG_OWNTC) {
+		assert(t->budget && t->period);
+
+		if (t->last_replenish == 0 || t->last_replenish + t->period <= now) {
+			tcap_res_t currbudget;
+
+			t->last_replenish = now;
+			currbudget        = (tcap_res_t)cos_introspect(ci, sl_thd_tcap(t), TCAP_GET_BUDGET);
```

Comment:
I'll add an issue and will remember to discuss with you on this. 

I don't think we have a good way to track the expenditure of a tcap. we need expenditure of a tcap especially in case of child component, but we literally cannot do anything about this right now. It gets complicated when we have the child component scheduler transfering time to its child AEPs. Even `introspect` is not a complete solution to this.